### PR TITLE
kademlia, hive: rate limiting on gossiping peers (send and receive)

### PR DESCRIPTION
**Pull request by esadakar**
#1645

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1654)

### CODE REVIEW

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -668,35 +672,39 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
...
+		}
+	}
 
 	if len(addrs) == 0 {
 		return nil
 	}
 
 	err := k.discovery.BroadcastPeers(ctx, peer, addrs...)
 	if err != nil {
+		fmt.Println("DISCONNECT")
```
Comment (anatollupacescu) [inline pkg/topology/kademlia/kademlia.go:707]: do we want to keep this line or was it intended to be something temp?

**Review by anatollupacescu — APPROVED** [review]

File: pkg/topology/kademlia/kademlia.go (hunk 1 above)
Comment (esadakar) [reply pkg/topology/kademlia/kademlia.go:707]: oh snap...

File: pkg/topology/kademlia/kademlia.go [hunk 2]
```diff
@@ -668,28 +672,31 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
 func (k *Kad) Announce(ctx context.Context, peer swarm.Address) error {
 	addrs := []swarm.Address{}
 
-	_ = k.connectedPeers.EachBinRev(func(connectedPeer swarm.Address, _ uint8) (bool, bool, error) {
-		if connectedPeer.Equal(peer) {
-			return false, false, nil
+	//r := rand.New(securerand.NewSource())
+
+	for bin := uint8(0); bin < swarm.MaxBins; bin++ {
```
Comment (mrekucci) [inline pkg/topology/kademlia/kademlia.go:677]: r: @acud is there a reason why `swarm.MaxBins` is a typed constant? If not, I'll remove the type and change the loop: `for bin := 0; bin < swarm.MaxBins; bin++ {...` (also in the affected code in the `pkg/swarm/proximity.go` file), more on this topic: https://blog.golang.org/constants

File: pkg/topology/kademlia/kademlia.go [hunk 3]
```diff
@@ -668,28 +672,31 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
 func (k *Kad) Announce(ctx context.Context, peer swarm.Address) error {
 	addrs := []swarm.Address{}
```
Comment (mrekucci) [inline pkg/topology/kademlia/kademlia.go:673]: r: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

File: pkg/topology/kademlia/kademlia.go [hunk 4]
```diff
@@ -1166,3 +1173,21 @@ func (k *Kad) Close() error {
 
 	return nil
 }
+
+func randomSubset(addrs []swarm.Address, count int) ([]swarm.Address, error) {
+
+	if count >= len(addrs) {
```
Comment (mrekucci) [inline pkg/topology/kademlia/kademlia.go:1179]: The addresses aren't shuffled when the count `count >= len(addrs)`, is this expected behavior?

File: pkg/topology/pslice/pslice.go [hunk 5]
```diff
@@ -95,6 +95,28 @@ func (s *PSlice) EachBinRev(pf topology.EachPeerFunc) error {
 	return nil
 }
 
+func (s *PSlice) BinPeers(bin uint8) []swarm.Address {
```
Comment (mrekucci) [inline pkg/topology/pslice/pslice.go:98]: r: if you remove the type from `swarm.MaxBin` constant as recommended above, then the `bin` can be simply an integer and the casting on `L:102` wouldn't be necessary

**Review by acud — CHANGES_REQUESTED** [review]
> Reviewed 7 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1654)* status: 7 of 8 files reviewed, 11 unresolved discussions (waiting on @esadakar, @mrekucci, and @zelig)

___
*[go.mod, line 69 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwSGqPE4D9KIA-klWR:-MZwSGqPE4D9KIA-klWS:b-65yhi5) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/go.mod#L69)):*
> ```XML
> 	golang.org/x/term v0.0.0-20201210144234-2321bbc49cbf
> 	golang.org/x/text v0.3.4 // indirect
> 	golang.org/x/time v0.0.0-20191024005414-555d28b269f0
> ```

this change is probably not needed and can be reverted
___
*[pkg/hive/hive.go, line 40 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwSNlF-hz2F3zFVLU3:-MZwSNlF-hz2F3zFVLU4:b2dfa5l) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L40)):*
> ```Go
> 
> var (
> 	ErrRateLimitExceed = errors.New("rate limit exceeded")
> ```

r: maybe `ErrRateLimit` or `ErrRateLimitExceeded`
___
*[pkg/hive/hive.go, line 41 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT2ZdDgxYqf9PWAyJ:-MZwT2ZdDgxYqf9PWAyK:b-3fpu5n) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L41)):*
> ```Go
> var (
> 	ErrRateLimitExceed = errors.New("rate limit exceeded")
> 	limitBurst         = 4 * int(swarm.MaxBins)
> ```

this is still quite a bit (128 in a minute, 7700 in an hour, 180K a day), so the address spam vector is still there. that being said, i'm not sure I have a better idea. maybe limiting to 4 peers in a bin in a minute?
___
*[pkg/hive/hive.go, line 201 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT_-990WgsfoUeXkQ:-MZwT_-990WgsfoUeXkR:b-xso627) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L201)):*
> ```Go
> 	limiter, ok := s.limiter[addr]
> 	if !ok {
> 		limiter = rate.NewLimiter(limitRate, limitBurst)
> ```

you add to the map but never clean up. this is a clear memory leak. you'd need to leverage the `DisconnectIn` and `DisconnectOut` protocol helpers to detect that the peer went away (or we decided to disconnect from it) in order to clean up the entry for the peer
___
*[pkg/topology/kademlia/kademlia.go, line 677 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZvmt2QAOHQrK3BKH9W-r1-677:-MZwU8Zu8F5h0ZbTzfDP:bnvudd6) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/kademlia/kademlia.go#L677)):*
<details><summary><i>Previously, mrekucci (Peter Mrekaj) wrote…</i></summary><blockquote>

r: @acud is there a reason why `swarm.MaxBins` is a typed constant? If not, I'll remove the type and change the loop: `for bin := 0; bin < swarm.MaxBins; bin++ {...` (also in the affected code in the `pkg/swarm/proximity.go` file), more on this topic: https://blog.golang.org/constants
</blockquote></details>

it is typed since everything which is a PO is intended to be a `uint8`. Going with the untyped approach might make daily life a tad bit easier but I think that having it typed provides good assurances and a strong consideration for developers to actually keep in mind the upper bound when actually writing code
___
*[pkg/topology/kademlia/kademlia_test.go, line 1244 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwV2nO12dQ3SS0-oQj:-MZwV2nO12dQ3SS0-oQk:b-5jego6) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/kademlia/kademlia_test.go#L1244)):*
> ```Go
> 	t.Helper()
> 
> 	timeout := time.After(5 * time.Second)
> ```

is the change needed?
___
*[pkg/topology/pslice/pslice_test.go, line 260 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwV7ga9HwG86n0W6Zq:-MZwV7ga9HwG86n0W6Zr:bfjzx4o) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/pslice/pslice_test.go#L260)):*
> ```Go
> 	} {
> 
> 		t.Run("", func(t *testing.T) {
> ```

test label missing. we usually add a short strong in the actual `tc` struct that has the string label which is used here
___
*[pkg/topology/pslice/pslice_test.go, line 284 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwVL6Z75JFvSAI7ua6:-MZwVL6Z75JFvSAI7ua7:b-s8b3v8) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/pslice/pslice_test.go#L284)):*
> ```Go
> 			bins := ps.BinPeers(uint8(len(tc.peersCount)))
> 			if bins != nil {
> 				t.Fatal("peers must be nil for out of bound bind")
> ```

out of bound bin?

**Review by zelig — APPROVED** [review]
LGTM, heed @mrekucci 's comments

File: pkg/topology/kademlia/kademlia.go (hunk 4 above)
Comment (esadakar) [reply pkg/topology/kademlia/kademlia.go:1179]: yes

**Review by esadakar — COMMENTED** [review]
> Dismissed @acud from 6 discussions.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1654)* status: 2 of 8 files reviewed, 6 unresolved discussions (waiting on @acud and @esadakar)

___
*[go.mod, line 69 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwSGqPE4D9KIA-klWR:-MZyRcjGENjTgEechQcy:b-4j7g4q) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/go.mod#L69)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

this change is probably not needed and can be reverted
</blockquote></details>

this gets added after go mod tidy
___
*[pkg/hive/hive.go, line 40 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwSNlF-hz2F3zFVLU3:-M_-ZLns9sCnfCI_11YS:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L40)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

r: maybe `ErrRateLimit` or `ErrRateLimitExceeded`
</blockquote></details>

Done.
___
*[pkg/hive/hive.go, line 41 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT2ZdDgxYqf9PWAyJ:-MZyMSnv536xlNekJCxY:bxrb0wn) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L41)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

this is still quite a bit (128 in a minute, 7700 in an hour, 180K a day), so the address spam vector is still there. that being said, i'm not sure I have a better idea. maybe limiting to 4 peers in a bin in a minute?
</blockquote></details>

I believe the way this rate limiting works is that we start with a token count B, capped at burst amount, and the rate R replenishes the token count. So at t=0, we have B tokens, so 128 are allowed in a single request. If 128 peers connect, than the token count falls to zero. After that, 1 peer per minute is allowed.  If no peers connect after the initial burst, then only after 128 minutes. the token count has replenished to the full burst amount. 
___
*[pkg/hive/hive.go, line 201 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT_-990WgsfoUeXkQ:-MZyO5dJ41KDd_xEk-eH:buk5m8e) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L201)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

you add to the map but never clean up. this is a clear memory leak. you'd need to leverage the `DisconnectIn` and `DisconnectOut` protocol helpers to detect that the peer went away (or we decided to disconnect from it) in order to clean up the entry for the peer
</blockquote></details>

Makes sense. My question would then be, what happens if a peer consistently reconnects, causing the rate limit counters to be wiped each time?  
___
*[pkg/topology/kademlia/kademlia_test.go, line 1244 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwV2nO12dQ3SS0-oQj:-MZyRG9bEHo4TbqcH1AA:b-83q4gn) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/kademlia/kademlia_test.go#L1244)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

is the change needed?
</blockquote></details>

No :)
___
*[pkg/topology/pslice/pslice_test.go, line 260 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwV7ga9HwG86n0W6Zq:-MZySYaL4pav4E8Cr0PI:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/pslice/pslice_test.go#L260)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

test label missing. we usually add a short strong in the actual `tc` struct that has the string label which is used here
</blockquote></details>

Done.
___
*[pkg/topology/pslice/pslice_test.go, line 284 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwVL6Z75JFvSAI7ua6:-MZySYB57B9mYWusnJ3A:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/topology/pslice/pslice_test.go#L284)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

out of bound bin?
</blockquote></details>

Done.

File: pkg/topology/kademlia/kademlia.go [hunk 6]
```diff
@@ -668,35 +672,39 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
 func (k *Kad) Announce(ctx context.Context, peer swarm.Address) error {
 	addrs := []swarm.Address{}
 
-	_ = k.connectedPeers.EachBinRev(func(connectedPeer swarm.Address, _ uint8) (bool, bool, error) {
-		if connectedPeer.Equal(peer) {
-			return false, false, nil
+	//r := rand.New(securerand.NewSource())
```
[the remark this answers is not in the document]
Comment (esadakar) [reply pkg/topology/kademlia/kademlia.go:675]: nope :)

**Review by acud — APPROVED** [review]
> Reviewed 1 of 4 files at r2, 5 of 5 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1654)* status: all files reviewed, 5 unresolved discussions (waiting on @esadakar)

___
*[pkg/hive/hive.go, line 201 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT_-990WgsfoUeXkQ:-M_-hc6GE96otRrfO-U5:b-kfrgfn) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L201)):*
<details><summary><i>Previously, esadakar (Esad Akar) wrote…</i></summary><blockquote>

Makes sense. My question would then be, what happens if a peer consistently reconnects, causing the rate limit counters to be wiped each time?  
</blockquote></details>

That is indeed possible :+1: nicely spotted. This is possible, but due to other nodes possible to be connecting in and reaching the maximum over saturated peer count in NN, meaning if they constantly do that they might get their slot taken by another node. But yeah this is indeed a thing. We should probably detect this sort of behavior in kademlia

**Review by acud — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1654#-:-M_0TuuPEh6VOBGraV2y:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @esadakar)

**Comment by ldeffenb** [comment]
___
*[pkg/hive/hive.go, line 41 at r1](https://reviewable.io/reviews/ethersphere/bee/1654#-MZwT2ZdDgxYqf9PWAyJ:-M_1NiVPDhiWDmnV8N2j:b-b0jhdg) ([raw file](https://github.com/ethersphere/bee/blob/a358f5720e893e66c13ecd099d1e7df1ece9b5c5/pkg/hive/hive.go#L41)):*
<details><summary><i>Previously, esadakar (Esad Akar) wrote…</i></summary><blockquote>

I believe the way this rate limiting works is that we start with a token count B, capped at burst amount, and the rate R replenishes the token count. So at t=0, we have B tokens, so 128 are allowed in a single request. If 128 peers connect, than the token count falls to zero. After that, 1 peer per minute is allowed.  If no peers connect after the initial burst, then only after 128 minutes. the token count has replenished to the full burst amount. 
</blockquote></details>

It's called a Token Bucket: https://en.wikipedia.org/wiki/Token_bucket

**Pull request merged**